src/Passes/Source/Commandline/ParameterParser.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Commandline/ParameterParser.hpp"
+
+#include <iostream>
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    void ParameterParser::parseArgs(int argc, char** argv)
+    {
+        uint64_t                 i = 1;
+        std::vector<ParsedValue> values;
+        while (i < static_cast<uint64_t>(argc))
+        {
+            values.push_back(parseSingleArg(argv[i]));
+            ++i;
+        }
+
+        i = 0;
+        while (i < values.size())
+        {
+            auto& v = values[i];
+            ++i;
+
+            if (!v.is_key)
+            {
+                arguments_.push_back(v.value);
+                continue;
+            }
+
+            if (i >= values.size())
+            {
+                settings_[v.value] = "true";
+                continue;
+            }
+
+            auto& v2 = values[i];
+            if (!v2.is_key && isOption(v.value))
+            {
+                settings_[v.value] = v2.value;
+                ++i;
+                continue;
+            }
+
+            settings_[v.value] = "true";
+        }
+    }
+
+    void ParameterParser::addFlag(String const& v)
+    {
+        flags_.insert(v);
+    }
+
+    ParameterParser::String ParameterParser::get(String const& name, String const& default_value) const noexcept
+    {
+        auto it = settings_.find(name);
+        if (it == settings_.end())
+        {
+            return default_value;
+        }
+
+        return it->second;
+    }
+
+    ParameterParser::String ParameterParser::get(String const& name) const
+    {
+        auto it = settings_.find(name);
+        if (it == settings_.end())
+        {
+            throw std::runtime_error("Could not find setting '" + name + "'.");
+        }
+
+        return it->second;
+    }
+
+    bool ParameterParser::has(String const& name) const noexcept
+    {
+        auto it = settings_.find(name);
+        return (it != settings_.end());
+    }
+
+    ParameterParser::Arguments const& ParameterParser::arguments() const
+    {
+        return arguments_;
+    }
+    ParameterParser::String const& ParameterParser::getArg(uint64_t const& n)
+    {
+        return arguments_[n];
+    }
```

Comment:
(Nit)
> `arguments_[n]`

This is a call to the `operator[]` of the `arguments_` container of type `std::vector<String>`. 
The type of the parameter (corresponding to the argument `n` above) of the `std::vector<String>::operator[]` is `std::vector<String>::size_type` (see [here](https://cplusplus.com/reference/vector/vector/operator[]/)).

You may want to consider the following change:
```suggestion
    ParameterParser::String const& ParameterParser::getArg(Arguments::size_type n)
    {
        return arguments_[n];
    }
```